In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Transaction_log_2019-10-05.csv')
df.head()

transaction_log_id                         trace_uid  merchant_id  \
0          3444288491  8aa984586cf68a00016d9a1301e00fbb         6489   
1          3444288492  8aa8852a6d8dbdb2016d9a13021d63d9        17650   
2          3444288493  8aa984556d3af82e016d9a13010950b5         1531   
3          3444288495  8aa984586c90d356016d9a1302575b7b         2430   
4          3444288498  8aa885486d8da314016d9a1303081c59         6489   

   location_id  terminal_number  product_id action_name balance_effect_type  \
0       250490     WALG04398T21       34667         Act                 ADD   
1       611221      vtsUS Games       22129         act                 ADD   
2       118657      CVS03177006       58021  PreAuthAct                NONE   
3       106891  WALWMT005390051       55193         Act                 ADD   
4       250490     WALG04398T21       34667         Act                 ADD   

   amount            transaction_date  ... transaction_log_id.1  alert_type  \
0    50.0  2019-10-05 00:00:00.060000  ...                  NaN         NaN   
1    10.0  2019-10-05 00:00:00.170000  ...                  NaN         NaN   
2    15.0  2019-10-05 00:00:00.350000  ...                  NaN         NaN   
3    10.0  2019-10-05 00:00:00.417000  ...                  NaN         NaN   
4    50.0  2019-10-05 00:00:00.420000  ...                  NaN         NaN   

  notification_type fraud_rule_id  version is_active.1  date_created.1  \
0               NaN           NaN      NaN         NaN             NaN   
1               NaN           NaN      NaN         NaN             NaN   
2               NaN           NaN      NaN         NaN             NaN   
3               NaN           NaN      NaN         NaN             NaN   
4               NaN           NaN      NaN         NaN             NaN   

   date_modified.1  threshold  fraud_rule_type  
0              NaN        NaN              NaN  
1              NaN        NaN              NaN  
2              NaN        NaN              NaN  
3              NaN        NaN              NaN  
4              NaN        NaN              NaN  

[5 rows x 36 columns]

In [3]:
print(df.shape)
df.describe()

(2859673, 36)


transaction_log_id   merchant_id   location_id    product_id  \
count        2.859673e+06  2.859673e+06  2.859673e+06  2.859673e+06   
mean         3.445717e+09  8.433359e+03  4.116088e+05  3.180148e+04   
std          8.251071e+05  6.847643e+03  3.003900e+05  1.825268e+04   
min          3.444288e+09  2.790000e+02  8.560000e+02  3.653000e+03   
25%          3.445003e+09  2.857000e+03  1.696320e+05  1.484100e+04   
50%          3.445717e+09  6.489000e+03  2.714400e+05  3.067100e+04   
75%          3.446432e+09  1.112900e+04  7.229200e+05  4.686800e+04   
max          3.447146e+09  2.535600e+04  1.064239e+06  6.150500e+04   

             amount  adjusted_amount           upc   category_id  \
count  2.859673e+06     2.859673e+06  2.859673e+06  2.858982e+06   
mean   2.879550e+02     2.655639e+02  8.237168e+11  5.052756e+00   
std    3.877242e+03     3.878469e+03  7.157768e+11  2.006392e+00   
min    0.000000e+00    -4.699000e+05  1.200000e+10  2.000000e+00   
25%    1.500000e+01     0.000000e+00  7.993662e+11  3.000000e+00   
50%    2.500000e+01     1.500000e+01  7.993665e+11  4.000000e+00   
75%    5.000000e+01     3.000000e+01  7.993667e+11  7.000000e+00   
max    5.000000e+05     5.000000e+05  8.858860e+12  2.100000e+01   

        terminal_id  pos_entry_type  payment_type  cashier_id  terminal_type  \
count  2.859673e+06    1.055149e+06           0.0         0.0            0.0   
mean   4.797451e+06    2.435842e+01           NaN         NaN            NaN   
std    2.822972e+06    1.075391e+01           NaN         NaN            NaN   
min    0.000000e+00    1.000000e+01           NaN         NaN            NaN   
25%    2.616269e+06    2.000000e+01           NaN         NaN            NaN   
50%    3.894059e+06    2.200000e+01           NaN         NaN            NaN   
75%    7.413984e+06    3.000000e+01           NaN         NaN            NaN   
max    9.432221e+06    9.000000e+02           NaN         NaN            NaN   

       alert_log_id  transaction_log_id.1  fraud_rule_id  version  \
count  4.935000e+03          4.935000e+03    4935.000000   4935.0   
mean   1.230303e+07          3.445619e+09  102158.132928      1.0   
std    1.424756e+03          9.741722e+05   32802.324411      0.0   
min    1.230056e+07          3.444289e+09    1242.000000      1.0   
25%    1.230180e+07          3.444645e+09  112531.000000      1.0   
50%    1.230303e+07          3.445594e+09  115672.000000      1.0   
75%    1.230426e+07          3.446535e+09  119269.000000      1.0   
max    1.230550e+07          3.447146e+09  119306.000000      1.0   

           threshold  
count    4935.000000  
mean      666.218237  
std      4608.396746  
min         1.000000  
25%       200.000000  
50%       250.000000  
75%       501.000000  
max    100001.000000

In [4]:
df.columns

Index(['transaction_log_id', 'trace_uid', 'merchant_id', 'location_id',
       'terminal_number', 'product_id', 'action_name', 'balance_effect_type',
       'amount', 'transaction_date', 'source_system', 'is_active',
       'date_created', 'date_modified', 'adjusted_amount', 'inventory_key',
       'upc', 'category_id', 'terminal_id', 'multipack_parent',
       'ignore_multipack_transaction', 'pos_entry_type', 'payment_type',
       'cashier_id', 'terminal_type', 'alert_log_id', 'transaction_log_id.1',
       'alert_type', 'notification_type', 'fraud_rule_id', 'version',
       'is_active.1', 'date_created.1', 'date_modified.1', 'threshold',
       'fraud_rule_type'],
      dtype='object')

In [5]:
null_cols = df.isna().mean().sort_values(ascending=False).reset_index()
null_cols

index         0
0                  terminal_type  1.000000
1                     cashier_id  1.000000
2                   payment_type  1.000000
3                fraud_rule_type  0.998274
4                      threshold  0.998274
5                date_modified.1  0.998274
6                 date_created.1  0.998274
7                    is_active.1  0.998274
8                        version  0.998274
9                  fraud_rule_id  0.998274
10             notification_type  0.998274
11                    alert_type  0.998274
12          transaction_log_id.1  0.998274
13                  alert_log_id  0.998274
14                pos_entry_type  0.631025
15                   category_id  0.000242
16              multipack_parent  0.000000
17                   terminal_id  0.000000
18                     trace_uid  0.000000
19                   merchant_id  0.000000
20                   location_id  0.000000
21               terminal_number  0.000000
22                    product_id  0.000000
23                   action_name  0.000000
24           balance_effect_type  0.000000
25                        amount  0.000000
26              transaction_date  0.000000
27                 source_system  0.000000
28                     is_active  0.000000
29                  date_created  0.000000
30                 date_modified  0.000000
31               adjusted_amount  0.000000
32                 inventory_key  0.000000
33                           upc  0.000000
34  ignore_multipack_transaction  0.000000
35            transaction_log_id  0.000000

In [6]:
"""transaction_log_id - The unique identifier to identify the transaction. no missing value. this is a unique number and the length of transaction_log_id = length of df.  
(Only used in GBM - In my opinion, we can't use this to create a model. We can use this for evaluation purpose)"""
print(df['transaction_log_id'].nunique())
print(df.shape)

2857674
(2859673, 36)


In [7]:
"""merchant_id - Unique Id represent a merchant. No missing values. There are 1029 unique values in our dataframe
In my opinion we can't use merchant_id for data modeling, we can use that for data cleaning or to create more data from the existing columns"""
print(df['merchant_id'].nunique())
print(df.shape)

1029
(2859673, 36)


In [8]:
"""location_id - Unique id to represent a location. No missing values, there are 137886 unique locations in our df.
Can't use this variable for data modeling, but we can make new data by combining this and other fields - (may be merchant_id and location_id to check if there is a hidden
pattern )
"""
print(df['location_id'].nunique())
print(df.shape)

137886
(2859673, 36)


In [9]:
"""4. product_id - Identifier to identify the product. No missing values. 7043 unique products. We may use this to identify which card/(s) are mostly used
location_id and product_id join together to see if more details may be on product or on fraudelent transactions(if any preference on 
particular product or location for fraudelent transactions. )
"""
print(df['product_id'].nunique())
print(df.shape)

7043
(2859673, 36)


In [10]:
"""5. terminal_number - Identify the terminal which is used. No missing values, 436733 unique terminal ids. Used in GBM, ANN and RNN
"""
print(df['terminal_number'].nunique())
print(df.shape)

436733
(2859673, 36)


In [11]:
"""6. action_name - No missing values, 32 unique values. We may have to do One-Hot-Encoding 
Not sure what is the relevance of these variables in ML model. 
"""
print(df['action_name'].nunique())
print(df.shape)
print(df['action_name'].unique())

32
(2859673, 36)
['Act' 'act' 'PreAuthAct' 'preauthact' 'Deact' 'valins' 'Issue'
 'prevalins' 'BalInq' 'Redeem' 'Sale' 'preauthrecharge' 'Recharge'
 'revvalins' 'ReverseAct' 'deact' 'reverseact' 'RedeemReversal' 'balinq'
 'SaleReversal' 'RedemptionAuthLock' 'RedemptionCompUnlock' 'StoreCredit'
 'sale' 'preauthESNAct' 'Return' 'baltransfer' 'RevDeact' 'RevValIns'
 'RevRedemptionAuthLock' 'return' 'StoreCreditReversal']


In [12]:
"""amount - No missing values, Mean value: 287.955, Max value: 500000.0 and Min value: 0
"""
print(df['amount'].mean())
print(df['amount'].max())
print(df['amount'].min())

287.9549781600905
500000.0
0.0


In [13]:
"""8. adjusted_amount - No missing values, Mean: 265.56, Max: 500000.0, min: -469900.0
"""
print(df['adjusted_amount'].mean())
print(df['adjusted_amount'].max())
print(df['adjusted_amount'].min())

265.56386956480696
500000.0
-469900.0


In [14]:
"""9. upc - Universal product code of the item. This is like an id. There are no missing items and 7043 unique values in our df.
This is exactly the same unique values of product_id. So upc and product_id might be the same. We need to check this two thorughly to understand the connection. 
"""
print(df['upc'].nunique())
print(df.shape)

7043
(2859673, 36)


In [15]:
"""10. category_id - This represents the category of the product. We have some missing values in this. There are 10 unique values in this dataframe. 
Most frequent value is 7. We can add that value to the missing fields. We may need to do One-Hot-Encoding for this too. 
"""
print(df['category_id'].nunique())
print(df['category_id'].mode())
print(df['category_id'].unique())
print(df.shape)

10
0    7.0
dtype: float64
[ 7.  3.  4.  8.  5.  2. 10. 21.  6. nan  9.]
(2859673, 36)


In [16]:
"""11. multipack_parent - Is the card a multipack parent(I think that shows if the card belongs to a multi pack card category  like 4x25).
No missing values. This has 2 unique values - True or False. So we may need to encode(binary encoding) the values. 
"""
print(df['multipack_parent'].nunique())
print(df['multipack_parent'].unique())
print(df.shape)

2
[ True False]
(2859673, 36)


In [17]:
print(df['ignore_multipack_transaction'].nunique())
print(df['ignore_multipack_transaction'].unique())

2
[ True False]


In [18]:
print(df['pos_entry_type'].nunique())
print(df['pos_entry_type'].mode())

11
0    22.0
dtype: float64


In [19]:
#feature extractions/feature generations
"""Adj_Amt_Location_15_Min - """

'Adj_Amt_Location_15_Min - '

Common Input features for models: There are 4 models -
1. GBM - Gradient boosting 
2. ANN - Artificial Neural Network
3. RNN - Recurrent Neural Network
4. LGBM - Light Gradient Boosting 
There are 18 columns common for all 4 algorithms. Most of them are columns in our dataframe, but there are some columns which are not in our df and is derived from the 
existing columns. My understanding and details of these columns:
1. transaction_log_id - The unique identifier to identify the transaction. There are no missing value. In this dataframe, this is a unique number and the length of transaction_log_id = length of df.  (Only used in GBM - In my opinion, we can't use this to create a model. We can use this for evaluation purpose)
2. merchant_id - Unique Id represent a merchant. No missing values. There are 1029 unique values in our dataframe
3. location_id - Unique id to represent a location. No missing values, there are 137886 unique locations in our df.
4. product_id - Identifier to identify the product. No missing values. 7043 unique products
5. terminal_number - Identify the terminal which is used. No missing values, 436733 unique terminal ids. Used in GBM, ANN and RNN
6. action_name - No missing values, 32 unique values. We may have to do One-Hot-Encoding 
7. amount - No missing values, Mean value: 287.955, Max value: 500000.0 and Min value: 0
8. adjusted_amount - No missing values, Mean: 265.56, Max: 500000.0, min: -469900.0
9. upc - Universal product code of the item. This is like an id. There are no missing items and 7043 unique values in our df. This is exactly the same unique values of product_id. So upc and product_id might be the same. We need to check this two thorughly to understand that. 
10. category_id - This represents the category of the product. We have some missing values in this. There are 10 unique values in this dataframe. Most frequent value is 7. We can add that value to the missing fields. We may need to do One-Hot-Encoding for this too. 
11. multipack_parent - Is the card a multipack parent(I think that shows if the card belongs to a multi pack card category  like 4x25). No missing values. This has 2 unique values - True or False. So we may need to encode(binary encoding) the values. 
12. ignore_multipack_transaction - This shows if the card belongs to a multi-pack or not. There are no missing values, only 2 options - True or False. We can impute values using simple imputer and drop_first=True
13. pos_entry_type - Entry type of point of sale, eg: manual. There are 11 unique items in this dataset and 63% data missing. (This is an ordinal data and the most frequent value is 22. (not sure what is the significance of 22). So we can fill the columns with the mode value.)
In my opinion, we shouldn't use this value for our model building because when more than 20-25% missing, the data may not have standard distribution and this will affect our modeling. 
14. transaction_date - no missing value. This is used to derieve more data. So this is a very important column. 
The transaction_date column includes date and time of transaction. In this df, all columns should have same date value since we are analyzing one particular day transaction. But have different value for the time. This may give more idea. We can extract details from that particular column like - 
    * what day of the week transaction occur?
    * Was that a holiday?
    * Was that a weekend?
    * Is there any pattern in fraudelent transactions? like AM or PM
    * Try to find hidden link between the location, time and fraudelent transactions
15. Adj_Amt_Location_15_Min - This is a derived column from adjusted_amount, location_id and transaction_date. In this column we derive a new column so that we can get the aggregate value of transactions happen in a location in 15 mins interval. 
(Not sure if this 15 mins before transaction or after transaction or interval 15 mins which include both before and after the transactions)
16. Adj_Amt_Location_30_Min - same as above, only difference is that instead of 15 mins we are checking 30 mins time. 
17. Adj_Amt_Location_1_Hour
18. Adj_Amt_Terminal_15_Min: In this case instead of location, we are trying to understand the adjusted amount happen in a terminal. (we are narrowing down the location to a particular terminal, this may give a precise information.)
Adj_Amt_Terminal_15_Min - eventhough this gives precise information, this may make lot more data/transactions. I am not sure how this is relevant when we (already) have location wise calculations. Especially if we don't have any particulars to identify if these transactions are by same customer or different customers(I think we have no way of knowing who made the transactions. If a customer buy 10 different gift cards in a single transaction,  we know it is done by same customer else if customer make different trnsactions, we can't even differentiate it) In that case is this relevant? 
19. Adj_Amt_Terminal_30_Min:
20. Adj_Amt_Terminal_1_Hour
21. day_of_transaction: I think this is also a derived column from transaction_date and in my opinion this is very important and relevant. (We may also need to find if it is a weekend or non-weekend. There would be 2 values and we may need to impute that for further calculation). At the end we may have to do One_Hot to encode values. 
22. Cnt_Adj_Amt_Location_15_Min: Thia is a derived column. In my opinion this is also relevant provided if we can track more details.
23. Cnt_Adj_Amt_Location_30_Min:
24. Cnt_Adj_Amt_Location_1_Hour:
Only one question regarding the derived column is why we are using adjusted amount, why not amount? 
25. Cnt_Adj_Amt_Terminal_15_Min
26. Cnt_Adj_Amt_Terminal_30_Min
27. Cnt_Adj_Amt_Terminal_1_Hour
Not sure, how we treat this value. This is an ordinal value and how are we going to enode? because the value may have different aggregate at different days and different times. So not sure how to take that for calculations